src/main/java/org/sagebionetworks/bridge/hibernate/MySQLHibernatePersistenceExceptionConverter.java

```diff
@@ -0,0 +1,109 @@
+    // And the SQL-specific error message indicates something about the type of constraint and thus, the 
+    // appropriate error message. We look for the following keys:
+    
+    /**
+     * Foreign key constraints: the name of the constraint is used to inform the caller about what object is 
+     * using the target entity. Constraints that cascade deletes do not need to be added to this map (updates
+     * seem to always be handled by Hibernate when they are needed).
+     */
+    private static final Map<String,String> FOREIGN_KEY_CONSTRAINTS = new ImmutableMap.Builder<String,String>()
+            .put("AssessmentRef-Assessment-Constraint", "a scheduling session")
+            .put("Schedule-Organization-Constraint", "a schedule")
+            .build();
+    
+    /**
+     * Unique key constraints: the name of the constraint is used to inform the caller about the fields that are 
+     * being duplicated.
+     */
+    private static final Map<String,String> UNIQUE_KEY_CONSTRAINTS = new ImmutableMap.Builder<String,String>()
+            .put("PRIMARY", "primary keys")
+            .put("TimeWindow-guid-sessionGuid-idx", "time window GUIDs")
+            .put("Session-guid-scheduleGuid-idx", "session GUIDs")
+            .build();
+
+    @Override
+    public RuntimeException convert(PersistenceException exception, Object entity) {
+        String name = BridgeUtils.getTypeName(entity.getClass());
+        
+        if (exception instanceof OptimisticLockException) {
+            return new ConcurrentModificationException(String.format(WRONG_VERSION_MSG, name));
+        }
+        if (exception instanceof NonUniqueObjectException) {
+            Serializable identifier = ((NonUniqueObjectException)exception).getIdentifier();
+            return new ConstraintViolationException.Builder()
+                    .withMessage(String.format(NON_UNIQUE_MSG, name.toLowerCase(), identifier)).build();
+        }
+
+        Throwable throwable = Throwables.getRootCause(exception);
+        if (throwable instanceof java.sql.SQLIntegrityConstraintViolationException) {
+            String rawMessage = throwable.getMessage();
+            String displayMessage = DEFAULT_CONSTRAINT_MSG;
+            
+            // SQLIntegrityConstraintViolation can contain different constraint violations, 
+            // including foreign key and duplicate entry violations
+            if (rawMessage.contains("Duplicate entry")) {
+                displayMessage = selectMsg(
+                        rawMessage, UNIQUE_KEY_CONSTRAINTS, UNIQUE_CONSTRAINT_MSG, name, displayMessage);
+            } else if (rawMessage.contains("a foreign key constraint fails")) {
+                displayMessage = selectMsg(
+                        rawMessage, FOREIGN_KEY_CONSTRAINTS, FK_CONSTRAINT_MSG, name, displayMessage);
+            }
+            return new ConstraintViolationException.Builder().withMessage(displayMessage).build();
+        }
+        return exception;
```

Comment:
HibernateAssessment needs a @BridgeTypeName annotation, and other objects might as well.